### ISSUE TITLE
Using a less advanced version of generic

### ISSUE THREAD

**Issue by hbagdi** [ent/ent#3463]
Hello there,

Thank you for this project, my experience so far has been very positive and it performs a lot of heavy lifting for me.

I'm using golangci-lint in my project and I've been facing an issue for a while now. The linter barfs on code generated after 0.11.4 with the following:

```
./bin/golangci-lint run ./...                                                                                                                                                                                                                                                                                                                                                                                                                                                              
ERRO [linters_context/goanalysis] buildssa: panic during analysis: in <snipped>/ent.withHooks$1: cannot convert *t0 (M) to PM, goroutine 20762 [running]:                                                                                                                                                                                                                                                                                                         
runtime/debug.Stack()                                                                                                                                                                                                                                                                                                                                                                                                                                                                      
        runtime/debug/stack.go:24 +0x65                                                                                                                                                                                                                                                                                                                                                                                                                                                    
github.com/golangci/golangci-lint/pkg/golinters/goanalysis.(*action).analyzeSafe.func1()                                                                                                                                                                                                                                                                                                                                                                                                   
        github.com/golangci/golangci-lint/pkg/golinters/goanalysis/runner_action.go:105 +0x6b                                                                                                                                                                                                                                                                                                                                                                                              
panic({0xf74060, 0xc00f3e9850})                                                                                                                                                                                                                                                                                                                                                                                                                                                            
        runtime/panic.go:884 +0x213                                                                                                                                                                                                                                                                                                                                                                                                                                                        
golang.org/x/tools/go/ssa.emitConv(0xc011497b00, {0x1362de0, 0xc00f9858c0}, {0x135ac78?, 0xc01d902870})                                                                                                                                                                                                                                                                                                                                                                                    
        golang.org/x/tools@v0.7.0/go/ssa/emit.go:286 +0xdbc                                                                                                                                                                                                                                                                                                                                                                                                                                
golang.org/x/tools/go/ssa.emitStore(0xc011497b00, {0x1362de0, 0xc00f985800}, {0x1362de0, 0xc00f9858c0}, 0x8d8d4bc)                                                                                                                                                                                                           
        golang.org/x/tools@v0.7.0/go/ssa/emit.go:377 +0x5f                                                                                                                                                                                                                                                                   
golang.org/x/tools/go/ssa.(*address).store(0xc00f3fb890, 0xc011497b00?, {0x1362de0?, 0xc00f9858c0?})                                                                                                                                                                                                                         
        golang.org/x/tools@v0.7.0/go/ssa/lvalue.go:40 +0x47                                                                                                                                                                                                                                                                  
golang.org/x/tools/go/ssa.(*storebuf).emit(...)                                                                                                                                                                                                                                                                              
        golang.org/x/tools@v0.7.0/go/ssa/builder.go:533                                                                                                                     
golang.org/x/tools/go/ssa.(*builder).assignStmt(0xc011497b00?, 0xc011497b00, {0xc008b803e0, 0x1, 0x10ac320?}, {0xc008b803f0, 0x1, 0xc002a367b0?}, 0x0)                      
        golang.org/x/tools@v0.7.0/go/ssa/builder.go:1207 +0x43d                                                                                                             
golang.org/x/tools/go/ssa.(*builder).stmt(0xc002a368e8?, 0xc011497b00, {0x135d690?, 0xc0117a0dc0?})                                                                         
        golang.org/x/tools@v0.7.0/go/ssa/builder.go:2181 +0x425                                                                                                             
golang.org/x/tools/go/ssa.(*builder).stmtList(0x1?, 0xc00f3fb680?, {0xc0117a0e00?, 0x4, 0xc000674000?})                                                                     
        golang.org/x/tools@v0.7.0/go/ssa/builder.go:946 +0x45                                                                                                               
golang.org/x/tools/go/ssa.(*builder).stmt(0xc011497b00?, 0xc011497b00, {0x135d7b0?, 0xc01a3ba1b0?})                                                                         
        golang.org/x/tools@v0.7.0/go/ssa/builder.go:2277 +0x859                                                                                                             
golang.org/x/tools/go/ssa.(*builder).buildFunctionBody(0x0?, 0xc011497b00)                                                                                                  
        golang.org/x/tools@v0.7.0/go/ssa/builder.go:2391 +0x437                                                                                                             
golang.org/x/tools/go/ssa.(*builder).expr0(0xc002a37a38, 0xc0052d7680, {0x135da50?, 0xc008b80420?}, {0x7, {0x135ab88, 0xc005123140}, {0x0, 0x0}})                           
        golang.org/x/tools@v0.7.0/go/ssa/builder.go:656 +0x4d6                                                                                                              
golang.org/x/tools/go/ssa.(*builder).expr(0xfb6f60?, 0xc0052d7680, {0x135da50, 0xc008b80420})                                                                               
        golang.org/x/tools@v0.7.0/go/ssa/builder.go:625 +0x17b                                                                                                              
golang.org/x/tools/go/ssa.(*builder).expr0(0xc002a37a38, 0xc0052d7680, {0x135d810?, 0xc0117a0e80?}, {0x7, {0x135ab10, 0xc00f9a3810}, {0x0, 0x0}})                           
        golang.org/x/tools@v0.7.0/go/ssa/builder.go:676 +0x705                                                                                                              
golang.org/x/tools/go/ssa.(*builder).expr(0xc00f985200?, 0xc0052d7680, {0x135d810, 0xc0117a0e80})                                                                           
        golang.org/x/tools@v0.7.0/go/ssa/builder.go:625 +0x17b                                                                                                              
golang.org/x/tools/go/ssa.(*builder).assign(0xc0052d7680?, 0xc0052d7680?, {0x135fe20?, 0xc00f3fb5f0}, {0x135d810?, 0xc0117a0e80?}, 0x8?, 0x0)                               
        golang.org/x/tools@v0.7.0/go/ssa/builder.go:598 +0x3db                                                                                                              
golang.org/x/tools/go/ssa.(*builder).localValueSpec(0xc0052d7680?, 0xc0052d7680, 0xc0175557c0)                                                                              
        golang.org/x/tools@v0.7.0/go/ssa/builder.go:1147 +0xe5                                                                                                              
golang.org/x/tools/go/ssa.(*builder).stmt(0xc002a37598?, 0xc0052d7680, {0x135d900?, 0xc008b80460?})                                                                         
        golang.org/x/tools@v0.7.0/go/ssa/builder.go:2147 +0x17c5                                                                                                            
golang.org/x/tools/go/ssa.(*builder).stmtList(0xc00f984f00?, 0xc002a375e8?, {0xc00bd77080?, 0x8, 0xc002a37638?})                                                                                                                             
        golang.org/x/tools@v0.7.0/go/ssa/builder.go:946 +0x45                                                                                                 
golang.org/x/tools/go/ssa.(*builder).stmt(0xc0052d7680?, 0xc0052d7680, {0x135d7b0?, 0xc01a3ba4e0?})                                                                                                                                          
        golang.org/x/tools@v0.7.0/go/ssa/builder.go:2277 +0x859                                                                                               
golang.org/x/tools/go/ssa.(*builder).buildFunctionBody(0x943d11?, 0xc0052d7680)                                                                                                                                                                                                                                              
        golang.org/x/tools@v0.7.0/go/ssa/builder.go:2391 +0x437                                                                                                                                                                                                                                                              
golang.org/x/tools/go/ssa.(*builder).buildFunction(0x943d60?, 0xc0052d7680)                                                                                                                                                                                                                                                  
        golang.org/x/tools@v0.7.0/go/ssa/builder.go:2326 +0x2e                                                                                                                                                                                                                                                               
golang.org/x/tools/go/ssa.(*builder).buildCreated(0xc002a37a38)                                                                                                                                                                                                                                                              
        golang.org/x/tools@v0.7.0/go/ssa/builder.go:2413 +0x25                                                                                                                                                                               
golang.org/x/tools/go/ssa.(*Package).build(0xc01a75d480)                                                                                                                                                                                                                                                                     
        golang.org/x/tools@v0.7.0/go/ssa/builder.go:2606 +0xc86                                                                                                                                                                                                                                                              
sync.(*Once).doSlow(0xc006f52180?, 0xc00932e140?)                                                                                                                                                                                                                                                                            
        sync/once.go:74 +0xc2                                                                                                                                                                                                                                                                                                
sync.(*Once).Do(...)                                                                                                                                                                                                                                                                                                         
        sync/once.go:65                                                                                                                                                                                                                                                                                                      
golang.org/x/tools/go/ssa.(*Package).Build(...)                                                                                                                                                                                                                                                                              
        golang.org/x/tools@v0.7.0/go/ssa/builder.go:2477                                                                                                                                                                                     
golang.org/x/tools/go/analysis/passes/buildssa.run(0xc01725ef00)                                                                                                                                                                             
        golang.org/x/tools@v0.7.0/go/analysis/passes/buildssa/buildssa.go:72 +0x1a8                                                                                                                                                                                                                                          
github.com/golangci/golangci-lint/pkg/golinters/goanalysis.(*action).analyze(0xc0024350f0)                                                                                                                                                                                                                                   
        github.com/golangci/golangci-lint/pkg/golinters/goanalysis/runner_action.go:195 +0xa25                                                                                                                                               
github.com/golangci/golangci-lint/pkg/golinters/goanalysis.(*action).analyzeSafe.func2()                                                                                                                                                     
        github.com/golangci/golangci-lint/pkg/golinters/goanalysis/runner_action.go:113 +0x1d                                                                                                                                                                                                                                
github.com/golangci/golangci-lint/pkg/timeutils.(*Stopwatch).TrackStage(0xc0019399a0, {0x110f6e2, 0x8}, 0xc00b4aff48)                                                                                                                                                                                                        
        github.com/golangci/golangci-lint/pkg/timeutils/stopwatch.go:111 +0x4a                                                                                                                                                                                                                                               
github.com/golangci/golangci-lint/pkg/golinters/goanalysis.(*action).analyzeSafe(0x0?)                                                                                                                                                                                                                                       
        github.com/golangci/golangci-lint/pkg/golinters/goanalysis/runner_action.go:112 +0x85                                                                                                                                                                                                                                
github.com/golangci/golangci-lint/pkg/golinters/goanalysis.(*loadingPackage).analyze.func2(0xc0024350f0)                                                                                                                                                                                                                     
        github.com/golangci/golangci-lint/pkg/golinters/goanalysis/runner_loadingpackage.go:80 +0xb4                                                                                                                                                                                                                         
created by github.com/golangci/golangci-lint/pkg/golinters/goanalysis.(*loadingPackage).analyze                                                                                                                                              
        github.com/golangci/golangci-lint/pkg/golinters/goanalysis/runner_loadingpackage.go:75 +0x208                                                                                                                                                                                                                        
ERRO [runner] Panic: bodyclose: package "ent" (isInitialPkg: true, needAnalyzeSource: true): interface conversion: interface {} is nil, not *buildssa.SSA: goroutine 20837 [running]:                                                                                                                                        
runtime/debug.Stack()                                                                                                                                                                                                                                                                                                        
        runtime/debug/stack.go:24 +0x65                                                                                                                                                                                                                                                                                      
github.com/golangci/golangci-lint/pkg/golinters/goanalysis.(*action).analyzeSafe.func1()                                                                                                                                                                                                                                     
        github.com/golangci/golangci-lint/pkg/golinters/goanalysis/runner_action.go:109 +0x285                                                                                                                                                                                                                               
panic({0xfd4880, 0xc0102f7fb0})                                                                                                                                                                                                                                                                                              
        runtime/panic.go:884 +0x213                                                                                                                                                                                                          
github.com/timakin/bodyclose/passes/bodyclose.runner.run({0xc0244ece10, {0x0, 0x0}, 0x0, {0x0, 0x0}, 0x0, 0x0}, 0xc0244ece10)                                                                                                                                                                                                
        github.com/timakin/bodyclose@v0.0.0-20221125081123-e39cf3fc478e/passes/bodyclose/bodyclose.go:45 +0x665                                                                                                                              
github.com/golangci/golangci-lint/pkg/golinters/goanalysis.(*action).analyze(0xc002435060)                                                                                                                                                   
        github.com/golangci/golangci-lint/pkg/golinters/goanalysis/runner_action.go:195 +0xa25                                                                                                                                               
github.com/golangci/golangci-lint/pkg/golinters/goanalysis.(*action).analyzeSafe.func2()                                                                                                                                                                                                                                     
        github.com/golangci/golangci-lint/pkg/golinters/goanalysis/runner_action.go:113 +0x1d                                                                                                                                                
github.com/golangci/golangci-lint/pkg/timeutils.(*Stopwatch).TrackStage(0xc0019399a0, {0x111e2a5, 0x9}, 0xc00300bf48)                                                                                                                        
        github.com/golangci/golangci-lint/pkg/timeutils/stopwatch.go:111 +0x4a                 
github.com/golangci/golangci-lint/pkg/golinters/goanalysis.(*action).analyzeSafe(0xc00300bfb0?)                                                                                                                                              
        github.com/golangci/golangci-lint/pkg/golinters/goanalysis/runner_action.go:112 +0x85                                                                                                                                                
github.com/golangci/golangci-lint/pkg/golinters/goanalysis.(*loadingPackage).analyze.func2(0xc002435060)                                                                                                                                     
        github.com/golangci/golangci-lint/pkg/golinters/goanalysis/runner_loadingpackage.go:80 +0xb4                                                                                                                                         
created by github.com/golangci/golangci-lint/pkg/golinters/goanalysis.(*loadingPackage).analyze                                                                                                                                              
        github.com/golangci/golangci-lint/pkg/golinters/goanalysis/runner_loadingpackage.go:75 +0x208                                                                                                                                        
WARN [runner] Can't run linter goanalysis_metalinter: goanalysis_metalinter: bodyclose: package "ent" (isInitialPkg: true, needAnalyzeSource: true): interface conversion: interface {} is nil, not *buildssa.SSA                            
ERRO Running error: 1 error occurred:                                                          
        * can't run linter goanalysis_metalinter: goanalysis_metalinter: bodyclose: package "ent" (isInitialPkg: true, needAnalyzeSource: true): interface conversion: interface {} is nil, not *buildssa.SSA                                
```


Is it possible to change the code generation to be use a less advanced form of generics to avoid this issue?
This really needs to be fixed in the linter itself (https://github.com/golangci/golangci-lint/issues/3711) and that is waiting on a fix to go's tooling (https://go-review.googlesource.com/c/tools/+/477635) which may or may not take a while.

**Comment by a8m** [comment]
Hey @hbagdi 👋 
https://github.com/ent/ent/pull/3360 (see comments there) - This patch was added to workaround this issue. Did you try updating your Ent version?

**Comment by hbagdi** [comment]
Doesn't that thread confirm that the problem still exists?

**Comment by a8m** [comment]
> Doesn't that thread confirm that the problem still exists?

It works for us in this repo and also internally, so I guess disabling some linters (as described [here](https://github.com/ent/ent/pull/3360#issuecomment-1475233582)) workaround the problem. cc @Antonboom to confirm I'm not wrong.

```
- noctx            # go 1.20, x/tools v0.7.0
- sqlclosecheck    # go 1.19, x/tools v0.5.0
- tparallel        # go 1.20, x/tools v0.7.0
- wastedassign     # go 1.14, x/tools v0.7.0
```

The reason I moved [`withHooks`](https://github.com/ent/ent/blob/master/entc/integration/ent/ent.go#L529-L560) to generics was to reduce the amount of generated code needed for each type+operation into a single function. Reverting this code means introducing again a lot of generated code which I prefer to avoid, but I'm open to any suggestions.

**Comment by Antonboom** [comment]
@a8m, hello!

The current state:
1) I trusted the [workaround](https://github.com/golang/go/issues/58633#issuecomment-1450447133) suggested by the Go team.

2) Due to circumstances unknown to me it worked at the time of local testing.

3) After the new release of the `golang-ci lint`, we turned out that this does not always work.

4) We found the following:
- [overlapping panics](https://github.com/golangci/golangci-lint/pull/3710) that interfere with testing and confuse
- saving the problem for a number of linters

5) I opened a separate issue, covered the workaround with tests in `golang/tools` repo and and it really didn't work. I proposed a solution to the Go team, which is awaiting review (this links presented above by topic-starter).

I planned to revert the **ent** changes (as we agreed) after the problem was completely gone for all linters.

I apologize for misleading, because I myself was not in the most pleasant situation.
